src/System.Text.Json/src/System/Text/Json/Serialization/Converters/DefaultIEnumerableConstructibleConveter.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Text.Json.Serialization.Policies;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class DefaultIEnumerableConstructibleConverter : JsonEnumerableConverter
+    {
+        // This method is invoked with reflection in CreateFromList.
+        private static IEnumerable CreateFromListInternal<TEnumerable, TElement>(IList sourceList)
+        {
+            List<TElement> list = new List<TElement>();
+
+            foreach (object item in sourceList)
```

Comment:
Use `for` to avoid the allocation on `GetEnumerator()`?

Reply:
Wouldn't that result in O(n^2) runtime since you end up indexing the `IList` using the indexer (whereas using the enumerator is O(n)?

Are you suggesting the trade-off between runtime vs allocations here is fine?

Reply:
Hopefully most IList have an O(1) indexer?

Reply:
What runtime type do we expect `sourceList` to be? And why are we using a type where indexing is O(N)? If that's true then maybe it makes sense to declare the parameter as `IEnumerable`.

I didn't imagine a tradeoff here because I assumed we'd use a backing type with O(1) indexing 😁

Reply:
> Hopefully most IList have an O(1) indexer?

Yes, good point. I hadn't realized that most (all?) of our IList types have O(1) indexers.

> I didn't imagine a tradeoff here because I assumed we'd use a backing type with O(1) indexing

OK. That makes more sense :)